### DISCUSSION TITLE
algorithm to relax mesh in areas with bad cell quality

### DISCUSSION THREAD

**Discussion by TermeHansen** [pyvista/pyvista#3511] [Ideas]
Hi all

I can't seem to find any post around on an algorithm to improve/relax a mesh in areas with bad cell quality. With the `DataSetFilters.compute_cell_quality()` I can find cells with for instance a high aspect ratio, but does there exist some kind of algorithm to try to relax the mesh by moving nodes along the mesh surface in a certain distance from the bad cell?

**Comment by darikg** [comment]
the [smooth](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.smooth.html#pyvista.PolyDataFilters.smooth) and [smooth_taubin](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.smooth_taubin.html#pyvista.PolyDataFilters.smooth_taubin) filters should both improve cell quality, but neither can be restricted to act specifically on a subset of cells. [vtkConstrainedSmoothingFilter](https://vtk.org/doc/nightly/html/classvtkConstrainedSmoothingFilter.html#details) could do that, though it's not wrapped in pyvista

**Reply by banesullivan** [reply]
You could perhaps split out the region with the `extract_cells` filter, smooth it, then rejoin the meshes. 

Good find, @darikg, we should definitely work to wrap the [`vtkConstrainedSmoothingFilter`](https://vtk.org/doc/nightly/html/classvtkConstrainedSmoothingFilter.html#details) algorithm
